pkg/querier/frontend2/scheduler.go

```diff
@@ -0,0 +1,510 @@
+func (s *Scheduler) cancelRequest(frontendAddr string, queryID uint64) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	key := requestKey{frontendAddr: frontendAddr, queryID: queryID}
+	req := s.pendingRequests[key]
+	if req != nil {
+		req.ctxCancel()
+	}
+	delete(s.pendingRequests, key)
+}
+
+// QuerierLoop is started by querier to receive queries from scheduler.
+func (s *Scheduler) QuerierLoop(querier SchedulerForQuerier_QuerierLoopServer) error {
+	resp, err := querier.Recv()
+	if err != nil {
+		return err
+	}
+
+	querierID := resp.GetQuerierID()
+
+	s.registerQuerierConnection(querierID)
+	defer s.unregisterQuerierConnection(querierID)
+
+	// If the downstream connection to querier is cancelled,
+	// we need to ping the condition variable to unblock getNextRequestForQuerier.
+	// Ideally we'd have ctx aware condition variables...
+	go func() {
+		<-querier.Context().Done()
+		s.cond.Broadcast()
+	}()
+
+	lastUserIndex := -1
+
+	// In stopping state scheduler is not accepting new queries, but still dispatching queries in the queues.
+	for s.isRunningOrStopping() {
+		req, idx, err := s.getNextRequestForQuerier(querier.Context(), lastUserIndex, querierID)
+		if err != nil {
+			return err
+		}
+		lastUserIndex = idx
+
+		if err := s.forwardRequestToQuerier(querier, req); err != nil {
+			return err
+		}
+	}
+
+	return errSchedulerIsNotRunning
+}
+
+func (s *Scheduler) forwardRequestToQuerier(querier SchedulerForQuerier_QuerierLoopServer, req *schedulerRequest) error {
+	// Make sure to cancel request at the end to cleanup resources.
+	defer s.cancelRequest(req.frontendAddress, req.queryID)
+
+	// Handle the stream sending & receiving on a goroutine so we can
+	// monitoring the contexts in a select and cancel things appropriately.
+	errCh := make(chan error, 1)
+	go func() {
+		err := querier.Send(&SchedulerToQuerier{
+			UserID:          req.userID,
+			QueryID:         req.queryID,
+			FrontendAddress: req.frontendAddress,
+			HttpRequest:     req.request,
+		})
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		_, err = querier.Recv()
+		errCh <- err
+	}()
+
+	select {
+	case <-req.ctx.Done():
+		// If the upstream request is cancelled (eg. frontend issued CANCEL or closed connection),
+		// we need to cancel the downstream req. Only way we can do that is to close the stream (by returning error here).
+		// Querier is expecting this semantics.
+		return req.ctx.Err()
+
+	case err := <-errCh:
+		// Is there was an error handling this request due to network IO,
+		// then error out this upstream request _and_ stream.
+		// TODO: if err is not nil, scheduler should notify frontend using the frontend address.
+		return err
+	}
+}
+
+// getQueue picks a random queue and takes the next unexpired request off of it, so we
+// fairly process users queries.  Will block if there are no requests.
+func (s *Scheduler) getNextRequestForQuerier(ctx context.Context, lastUserIndex int, querierID string) (*schedulerRequest, int, error) {
```

Comment:
This is also a copy of `Frontend.getNextRequestForQuerier()`

Reply:
Yes, I plan to refactor and unify this in follow-up PR, if that's fine. Idea is that frontend V1 will basically embed scheduler (hopefully while preserving its own metrics).

Reply:
Ok, let's do it in a separate PR.